File: packages/warlog-spec-ts/src/pack-manifest.ts
/**
 * PackManifest — the spec contract for distributing playbooks,
 * detection rules, and connector configs as signed registry packs.
 * Mirrors `warlog_spec.pack_manifest`.
 */

import { z } from "zod";

export const PACK_MANIFEST_VERSION = "1.0";

export const PackKind = z.enum([
  "playbook",
  "detection_rules",
  "kb_articles",
  "connector_spec",
  "action_mappings",
]);
export type PackKindValue = z.infer<typeof PackKind>;

export const TrustLevel = z.enum(["community", "verified", "certified"]);
export type TrustLevelValue = z.infer<typeof TrustLevel>;

export const PackPublisher = z.object({
  id: z.string().min(1).max(128),
  trustLevel: TrustLevel,
  signature: z.string(),
});
export type PackPublisher = z.infer<typeof PackPublisher>;

export const PackDependency = z.object({
  id: z.string(),
  version: z.string(),
});
export type PackDependency = z.infer<typeof PackDependency>;

export const PackCompat = z.object({
  warlogSpecMin: z.string(),
  warlogSpecMax: z.string(),
  dependsOnPacks: z.array(PackDependency).default([]),
});
export type PackCompat = z.infer<typeof PackCompat>;

export const PackContents = z.object({
  detectionRules: z.array(z.string()).default([]),
  playbooks: z.array(z.string()).default([]),
  kbArticles: z.array(z.string()).default([]),
  connectorSpecs: z.array(z.string()).default([]),
  actionMappings: z.array(z.string()).default([]),
  examples: z.array(z.string()).default([]),
  tests: z.array(z.string()).default([]),
});
export type PackContents = z.infer<typeof PackContents>;

export const PackProvenance = z.object({
  sourceRepo: z.string(),
  sourceCommit: z.string(),
  buildAt: z.string().datetime({ offset: true }),
  sbom: z.string().nullable().default(null),
  builderIdentity: z.string().nullable().default(null),
});
export type PackProvenance = z.infer<typeof PackProvenance>;

export const PackManifest = z.object({
  specVersion: z.literal("1.0").default("1.0"),
  packId: z.string().min(1).max(128),
  packVersion: z.string(),
  kind: PackKind,
  publisher: PackPublisher,
  title: z.string().min(1).max(200),
  description: z.string().min(1),
  compat: PackCompat,
  license: z.string(),
  contents: PackContents.default({
    detectionRules: [],
    playbooks: [],
    kbArticles: [],
    connectorSpecs: [],
    actionMappings: [],
    examples: [],
    tests: [],
  }),
  provenance: PackProvenance,
});
export type PackManifest = z.infer<typeof PackManifest>;

File: packages/warlog-spec-py/src/warlog_spec/pack_manifest.py
"""Pack manifest — distribution layer for the registry.

A pack is the unit of distribution in the Warlog registry: a versioned,
signed bundle of detection rules, playbooks, KB articles, connectors,
or response action mappings.
"""

from __future__ import annotations

from datetime import datetime
from enum import StrEnum
from typing import Literal

from pydantic import Field

from warlog_spec._base import SpecModel

PACK_MANIFEST_VERSION = "1.0"


class PackKind(StrEnum):
    """Five canonical pack categories."""

    DETECTION = "detection"
    PLAYBOOK = "playbook"
    KB = "kb"
    CONNECTOR = "connector"
    ACTION = "action"


class TrustLevel(StrEnum):
    """Publisher trust gradient — informs default install posture per
    tenant policy."""

    CERTIFIED = "certified"
    COMMUNITY = "community"
    PRIVATE = "private"


class PackPublisher(SpecModel):
    """Publisher identity and signature material."""

    id: str = Field(min_length=1, max_length=128)
    trust_level: TrustLevel
    signature: str


class PackDependency(SpecModel):
    """Reference to another pack this one needs."""

    id: str
    version: str


class PackCompat(SpecModel):
    """Spec + dependency compat declaration."""

    warlog_spec_min: str
    warlog_spec_max: str
    depends_on_packs: list[PackDependency] = Field(default_factory=list)


class PackContents(SpecModel):
    """Manifest of files inside the pack archive."""

    detection_rules: list[str] = Field(default_factory=list)
    playbooks: list[str] = Field(default_factory=list)
    kb_articles: list[str] = Field(default_factory=list)
    connector_specs: list[str] = Field(default_factory=list)
    action_mappings: list[str] = Field(default_factory=list)
    examples: list[str] = Field(default_factory=list)
    tests: list[str] = Field(default_factory=list)


class PackProvenance(SpecModel):
    """Supply-chain attestation for the pack."""

    source_repo: str
    source_commit: str
    build_at: datetime
    sbom: str | None = None
    builder_identity: str | None = None


class PackManifest(SpecModel):
    """Top-level manifest declared at the root of every pack archive."""

    spec_version: Literal["1.0"] = PACK_MANIFEST_VERSION
    pack_id: str = Field(min_length=1, max_length=128)
    pack_version: str
    kind: PackKind
    publisher: PackPublisher
    title: str = Field(min_length=1, max_length=200)
    description: str = Field(min_length=1)
    compat: PackCompat
    license: str
    contents: PackContents = Field(default_factory=PackContents)
    provenance: PackProvenance


__all__ = [
    "PACK_MANIFEST_VERSION",
    "PackCompat",
    "PackContents",
    "PackDependency",
    "PackKind",
    "PackManifest",
    "PackProvenance",
    "PackPublisher",
    "TrustLevel",
]
